executor/distsql.go

```diff
@@ -361,25 +366,78 @@ func (e *TableReaderExecutor) Open(goCtx goctx.Context) error {
 	span, goCtx := startSpanFollowsContext(goCtx, "executor.TableReader.Open")
 	defer span.Finish()
 
+	e.resultHandler = &tableResultHandler{}
+	firstPartRanges, secondPartRanges := e.splitRanges()
+	firstResult, err := e.buildResp(goCtx, firstPartRanges)
+	if err != nil {
+		e.feedback.Invalidate()
+		return errors.Trace(err)
+	}
+	if len(secondPartRanges) == 0 {
+		e.resultHandler.open(nil, firstResult)
+		return nil
+	}
+	var secondResult distsql.SelectResult
+	secondResult, err = e.buildResp(goCtx, secondPartRanges)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.resultHandler.open(firstResult, secondResult)
+	return nil
+}
+
+func (e *TableReaderExecutor) buildResp(goCtx goctx.Context, ranges []*ranger.NewRange) (distsql.SelectResult, error) {
 	var builder requestBuilder
-	kvReq, err := builder.SetTableRanges(e.tableID, e.ranges).
+	kvReq, err := builder.SetTableRanges(e.tableID, ranges).
 		SetDAGRequest(e.dagPB).
 		SetDesc(e.desc).
 		SetKeepOrder(e.keepOrder).
 		SetPriority(e.priority).
 		SetFromSessionVars(e.ctx.GetSessionVars()).
 		Build()
 	if err != nil {
-		e.feedback.Invalidate()
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	e.result, err = distsql.SelectDAG(goCtx, e.ctx, kvReq, e.retTypes())
+	result, err := distsql.SelectDAG(goCtx, e.ctx, kvReq, e.retTypes())
 	if err != nil {
-		e.feedback.Invalidate()
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	e.result.Fetch(goCtx)
-	return nil
+	result.Fetch(goCtx)
+	return result, nil
+}
+
+func (e *TableReaderExecutor) splitRanges() ([]*ranger.NewRange, []*ranger.NewRange) {
+	if len(e.ranges) == 0 || e.ranges[0].LowVal[0].Kind() == types.KindInt64 {
+		return e.ranges, nil
+	}
+	idx := sort.Search(len(e.ranges), func(i int) bool { return e.ranges[i].HighVal[0].GetUint64() > math.MaxInt64 })
+	if idx == len(e.ranges) {
+		return e.ranges, nil
+	}
+	signedRanges := make([]*ranger.NewRange, 0, idx+1)
+	unsignedRanges := make([]*ranger.NewRange, 0, len(e.ranges)-idx)
+	signedRanges = append(signedRanges, e.ranges[0:idx]...)
+	if e.ranges[idx].LowVal[0].GetUint64() <= math.MaxInt64 {
+		signedRanges = append(signedRanges, &ranger.NewRange{
+			LowVal:     e.ranges[idx].LowVal,
+			LowExclude: e.ranges[idx].LowExclude,
+			HighVal:    []types.Datum{types.NewIntDatum(math.MaxInt64)},
+		})
+		unsignedRanges = append(unsignedRanges, &ranger.NewRange{
+			LowVal:      []types.Datum{types.NewIntDatum(math.MinInt64)},
+			HighVal:     e.ranges[idx].HighVal,
+			HighExclude: e.ranges[idx].HighExclude,
+		})
+	} else {
+		unsignedRanges = append(unsignedRanges, e.ranges[idx])
```

Comment:
In this case, we don't need to make new ranges, we can just slice the original ranges.